src/components/timeline/Selection.js

```diff
@@ -0,0 +1,388 @@
+      // Do not stopPropagation(), so that graph gets mouseup event.
+      this._uninstallMoveAndUpHandlers();
+    };
+
+    this._installMoveAndUpHandlers(mouseMoveHandler, mouseUpHandler);
+  };
+
+  _installMoveAndUpHandlers(
+    mouseMoveHandler: MouseHandler,
+    mouseUpHandler: MouseHandler
+  ) {
+    this._handlers = { mouseMoveHandler, mouseUpHandler };
+    window.addEventListener('mousemove', mouseMoveHandler, true);
+    window.addEventListener('mouseup', mouseUpHandler, true);
+  }
+
+  _uninstallMoveAndUpHandlers() {
+    if (this._handlers) {
+      const { mouseMoveHandler, mouseUpHandler } = this._handlers;
+      window.removeEventListener('mousemove', mouseMoveHandler, true);
+      window.removeEventListener('mouseup', mouseUpHandler, true);
+    }
+  }
+
+  _onMouseMove = (event: SyntheticMouseEvent<>) => {
+    if (!this._container) {
+      return;
+    }
+
+    const rect = getContentRect(this._container);
+    if (
+      event.pageX < rect.left ||
+      event.pageX >= rect.right ||
+      event.pageY < rect.top ||
+      event.pageY >= rect.bottom
+    ) {
+      this.setState({ hoverLocation: null });
+    } else {
+      this.setState({ hoverLocation: event.pageX - rect.left });
+    }
+  };
+
+  _makeOnMove = (fun: number => { startDelta: number, endDelta: number }) => (
+    originalSelection: { +selectionStart: number, +selectionEnd: number },
+    dx: number,
+    dy: number,
+    isModifying: boolean
+  ) => {
+    const { displayRange, width, updateProfileSelection } = this.props;
+    const delta = dx / width * (displayRange.end - displayRange.start);
+    const selectionDeltas = fun(delta);
+    const selectionStart = Math.max(
+      displayRange.start,
+      originalSelection.selectionStart + selectionDeltas.startDelta
+    );
+    const selectionEnd = clamp(
+      originalSelection.selectionEnd + selectionDeltas.endDelta,
+      selectionStart,
+      displayRange.end
+    );
+    updateProfileSelection({
+      hasSelection: true,
+      isModifying,
+      selectionStart,
+      selectionEnd,
+    });
+  };
+
+  _rangeStartOnMove = this._makeOnMove(delta => ({
+    startDelta: delta,
+    endDelta: 0,
+  }));
+
+  _moveRangeOnMove = this._makeOnMove(delta => ({
+    startDelta: delta,
+    endDelta: delta,
+  }));
+
+  _rangeEndOnMove = this._makeOnMove(delta => ({
+    startDelta: 0,
+    endDelta: delta,
+  }));
+
+  _zoomButtonOnMouseDown = (e: SyntheticMouseEvent<>) => {
+    e.stopPropagation();
+  };
+
+  _zoomButtonOnClick = (e: SyntheticMouseEvent<>) => {
+    e.stopPropagation();
+    const { selection, zeroAt, addRangeFilterAndUnsetSelection } = this.props;
+    if (selection.hasSelection) {
+      addRangeFilterAndUnsetSelection(
+        selection.selectionStart - zeroAt,
+        selection.selectionEnd - zeroAt
+      );
+    }
+  };
+
+  renderSelectionOverlay() {
+    const { displayRange, selection, width } = this.props;
+    if (!selection.hasSelection) {
+      return null;
+    }
```

Comment:
I know it's a bit subjective: I think it's easier to read if the test is in the `render` function instead of here, because then with just the `render` function we can have a sense of what will actually be rendered.

Reply:
Sure, I can refine the selection type in the argument parameter.